File: 74-search-a-2d-matrix/74-search-a-2d-matrix.java
class Solution {
    // using binary search
    public boolean searchMatrix(int[][] matrix, int target) {
        int n = matrix.length, m = matrix[0].length;
        
        for(int i=0; i<n; i++) {
            int left = 0, right = m-1;
            
            if(target >= matrix[i][left] && target <= matrix[i][right]) {
                while(left<=right) {
                    int mid = left + (right-left) / 2;
                    
                    if(matrix[i][mid] == target) {
                        return true;
                    }else if(target > matrix[i][mid]) {
                        left = mid+1;
                    }else {
                        right = mid-1;
                    }
                }
                break;
            }
        }
        
        return false;
    }
}

File: 74-search-a-2d-matrix/74-search-a-2d-matrix.cpp
class Solution {
public:
    // the brute force
//     bool searchMatrix(vector<vector<int>>& matrix, int target) {
//         int found = 0;
//         for(auto i: matrix) {
//             for(auto j: i) {
//                 if(j==target) {
//                     found = 1; break;
//                 }
//             }
//         }
        
//         return found;
//     }
    
    // better optimised, binary search
    // Time complexity: O(Log m + Log n)
    // Space complexity: O(1)

    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int n = matrix.size(), m = matrix[0].size();
        int lower = 0, higher = n*m-1;
        
        while(lower<=higher) {
            int mid = lower + (higher - lower)/2;
            
            int val = matrix[mid/m][mid%m];
            
            if(val > target) {
                higher = mid-1;
            } else if(val < target) {
                lower = mid+1;
            } else {
                return true;
            }
        }
        
        return false;
    }
};